### PR TITLE
added convert command implementation

### PR DESCRIPTION
**Pull request by randrii**

### CODE REVIEW

File: src/main/java/com/rybka/command/ConvertCommand.java [hunk 1]
```diff
@@ -0,0 +1,16 @@
...
+
+import lombok.AllArgsConstructor;
+
+@AllArgsConstructor
+public class ConvertCommand implements Command {
+    private final ConvertTask convertTask;
+
+    @Override
+    public void execute() {
+        var convertedResult = convertTask.convert();
```
Comment (jarc0der) [inline src/main/java/com/rybka/command/ConvertCommand.java]: Why do you need ConvertTask then? 

put all logic and dependencies in the ConverCommand. 

In result you will have

ConvertCommand
ExportCommand
BlaBlaCommand

File: src/main/java/com/rybka/command/Invoker.java [hunk 2]
```diff
@@ -0,0 +1,12 @@
+package com.rybka.command;
+
+import lombok.AllArgsConstructor;
+
+@AllArgsConstructor
+public class Invoker {
```
Comment (jarc0der) [inline src/main/java/com/rybka/command/Invoker.java]: Can't find this requirement in the task description. Point me please where you find that service to invoke command should be implemented. 🥇

File: src/main/java/com/rybka/view/ExchangeView.java [hunk 3]
```diff
@@ -34,18 +37,10 @@ public void showDialog() {
...
-            var convertedResult = constructConvertedResult(currencyResponse.getBase(),
-                    currencyResponse.getTarget(),
-                    userValue,
-                    currencyResponse.getRate(),
-                    total);
-
-            currencyHistoryDAO.save(convertedResult);
-
-            showExchange(convertedResult);
+            var convertTask = new ConvertTask(service, currencyHistoryDAO, userBaseCurrency, userTargetCurrency, userValue);
```
Comment (jarc0der) [inline src/main/java/com/rybka/view/ExchangeView.java]: Get rid of ConvertTask, move all logic to ConvertCommand.

File: src/main/java/com/rybka/command/ConvertTask.java [hunk 4]
```diff
@@ -0,0 +1,46 @@
...
+import com.rybka.service.exchange.ExchangeService;
+import lombok.AllArgsConstructor;
+import lombok.extern.log4j.Log4j;
+
+@Log4j
+@AllArgsConstructor
+public class ConvertTask {
+    private final ExchangeService service;
+    private final CurrencyHistoryDAO currencyHistoryDAO;
+    private final String userBaseCurrency;
```
Comment (jarc0der) [inline src/main/java/com/rybka/command/ConvertTask.java]: How are you going to provide this values?

File: src/main/java/com/rybka/view/ExchangeView.java (hunk 3 above)
Comment (jarc0der) [inline src/main/java/com/rybka/view/ExchangeView.java]: Each time showDialog() method called we will build new ConvertCommand and execute it. Do you think this is correct? 
Would be better to define service and reuse it each time when command called?

File: src/main/java/com/rybka/command/ConvertCommand.java (hunk 1 above)
Comment (randrii) [reply src/main/java/com/rybka/command/ConvertCommand.java]: fixed

File: src/main/java/com/rybka/command/Invoker.java (hunk 2 above)
Comment (randrii) [reply src/main/java/com/rybka/command/Invoker.java]: fixed

File: src/main/java/com/rybka/view/ExchangeView.java (hunk 3 above)
Comment (randrii) [reply src/main/java/com/rybka/view/ExchangeView.java]: fixed

File: src/main/java/com/rybka/command/ConvertTask.java (hunk 4 above)
Comment (randrii) [reply src/main/java/com/rybka/command/ConvertTask.java]: fixed

File: src/main/java/com/rybka/command/ConvertCommand.java [hunk 5]
```diff
@@ -0,0 +1,77 @@
...
+
+@Log4j
+@RequiredArgsConstructor
+public class ConvertCommand implements Command {
+    private final Scanner scanner;
+    private final ExchangeService service;
+    private final CurrencyHistoryDAO currencyHistoryDAO;
+    private final ExportService exportService;
+
+    private String userBaseCurrency;
```
Comment (jarc0der) [inline src/main/java/com/rybka/command/ConvertCommand.java]: Why do you keep this variable here again????

File: src/main/java/com/rybka/command/ConvertCommand.java [hunk 6]
```diff
@@ -0,0 +1,77 @@
...
+    @Override
+    public void execute() {
+        try {
+            System.out.print("Insert base currency: ");
+            userBaseCurrency = scanner.next().toUpperCase();
+            System.out.print("Enter value: ");
+            userValue = scanner.nextDouble();
+            System.out.print("Insert target currency: ");
+            userTargetCurrency = scanner.next().toUpperCase();
+
```
Comment (jarc0der) [inline src/main/java/com/rybka/command/ConvertCommand.java:27]: 1. Create method `readUserParameters()` -> UserConvertData

UserConvertData
 + userBaseCurrency
 + userValue
 + userTargetCurrency

2. Call methods to convert currency 

var currencyResponse = service.loadCurrencyOf(userData.getUserBaseCurrency, userData.getUserTargetCurrency);
var total = service.calculateTotal(currencyResponse.getRate(), userData.getUserValue);

3. Save result to DB(history)
4. Print result on the console.

File: src/main/java/com/rybka/command/ConvertCommand.java [hunk 7]
```diff
@@ -0,0 +1,77 @@
...
+            var convertedResult = convert();
+            print(convertedResult);
+            save(convertedResult);
+
+            exportService.export(currencyHistoryDAO.findAll());
```
Comment (jarc0der) [inline src/main/java/com/rybka/command/ConvertCommand.java]: Why do you do export here? It's convert command.

File: src/main/java/com/rybka/command/ConvertCommand.java [hunk 8]
```diff
@@ -0,0 +1,77 @@
...
+        } catch (NullPointerException exception) {
+            log.error("Invalid currency for exchange.");
```
Comment (jarc0der) [inline src/main/java/com/rybka/command/ConvertCommand.java]: Are you sure NPE means invalidCurrency? 
What if currencyHistoryDAO = null? We will get the same message.

File: src/main/java/com/rybka/command/ConvertCommand.java [hunk 9]
```diff
@@ -0,0 +1,77 @@
...
+        } catch (InputMismatchException exception) {
+            log.error("Please enter valid value for exchange.");
+        } catch (Exception exception) {
+            log.error("Some issues are occurred! Reason: " + exception.getMessage());
+        }
+    }
+
+    private CurrencyHistory convert() {
```
Comment (jarc0der) [inline src/main/java/com/rybka/command/ConvertCommand.java]: If you use method, put parameters into it.

File: src/main/java/com/rybka/Application.java [hunk 10]
```diff
@@ -50,14 +51,16 @@ public static void main(String[] args) {
...
         ExchangeView view = new ExchangeView(
-                new Scanner(System.in),
-                new ExchangeService(connector),
-                new CurrencyHistoryDAO(),
-                exportService
+                new ConvertCommand(
+                        new Scanner(System.in),
+                        new ExchangeService(connector),
+                        new CurrencyHistoryDAO(),
+                        exportService
```
Comment (jarc0der) [inline src/main/java/com/rybka/Application.java]: You don't need export service here. Or I'm wrong?

File: src/main/java/com/rybka/command/ConvertCommand.java (hunk 6 above)
Comment (randrii) [reply src/main/java/com/rybka/command/ConvertCommand.java:27]: fixed

File: src/main/java/com/rybka/command/ConvertCommand.java (hunk 5 above)
Comment (randrii) [reply src/main/java/com/rybka/command/ConvertCommand.java]: fixed

File: src/main/java/com/rybka/command/ConvertCommand.java (hunk 7 above)
Comment (randrii) [reply src/main/java/com/rybka/command/ConvertCommand.java]: fixed

File: src/main/java/com/rybka/command/ConvertCommand.java (hunk 9 above)
Comment (randrii) [reply src/main/java/com/rybka/command/ConvertCommand.java]: fixed

File: src/main/java/com/rybka/Application.java (hunk 10 above)
Comment (randrii) [reply src/main/java/com/rybka/Application.java]: fixed

File: src/main/java/com/rybka/view/ExchangeView.java (hunk 3 above)
Comment (randrii) [reply src/main/java/com/rybka/view/ExchangeView.java]: fixed

File: src/main/java/com/rybka/command/ConvertCommand.java (hunk 8 above)
Comment (randrii) [reply src/main/java/com/rybka/command/ConvertCommand.java]: fixed

File: src/main/java/com/rybka/command/ConvertCommand.java [hunk 11]
```diff
@@ -0,0 +1,79 @@
...
+@RequiredArgsConstructor
+public class ConvertCommand implements Command {
+    private final Scanner scanner;
+    private final ExchangeService service;
+    private final CurrencyHistoryDAO currencyHistoryDAO;
+
+    @Override
+    public void execute() {
+        try {
+            System.out.print("Insert base currency: ");
```
Comment (jarc0der) [inline src/main/java/com/rybka/command/ConvertCommand.java]: Better to move this logic to `readUserParameters` method.

File: src/main/java/com/rybka/command/ConvertCommand.java [hunk 12]
```diff
@@ -0,0 +1,79 @@
...
+            save(convertedResult);
+        } catch (InputMismatchException exception) {
+            log.error("Please enter valid value for exchange.");
+        } catch (Exception exception) {
+            log.error("Some issues are occurred! Reason: " + exception.getMessage());
+        }
+    }
+
+    private CurrencyHistory convert(UserConvertData userConvertData) {
+        if (service == null) throw new CurrencyAPICallException("Unable to call exchange service.");
```
Comment (jarc0der) [inline src/main/java/com/rybka/command/ConvertCommand.java]: remove this check

File: src/main/java/com/rybka/command/ConvertCommand.java [hunk 13]
```diff
@@ -0,0 +1,79 @@
...
+                currencyResponse.getRate(),
+                total);
+    }
+
+    private void print(CurrencyHistory currencyHistory) {
+        showExchange(currencyHistory);
+    }
+
+    private void save(CurrencyHistory currencyHistory) {
+        if (currencyHistoryDAO == null) throw new DBConnectionException("Unable to connect to Database.");
```
Comment (jarc0der) [inline src/main/java/com/rybka/command/ConvertCommand.java]: remove null check

File: src/main/java/com/rybka/command/ConvertCommand.java (hunk 11 above)
Comment (randrii) [reply src/main/java/com/rybka/command/ConvertCommand.java]: fixed

File: src/main/java/com/rybka/command/ConvertCommand.java (hunk 12 above)
Comment (randrii) [reply src/main/java/com/rybka/command/ConvertCommand.java]: fixed

File: src/main/java/com/rybka/command/ConvertCommand.java (hunk 13 above)
Comment (randrii) [reply src/main/java/com/rybka/command/ConvertCommand.java]: fixed

**Pull request merged**